bytecomp/translcore.ml

```diff
@@ -1234,8 +1234,10 @@ and transl_setinstvar self var expr =
 
 and transl_record env all_labels repres lbl_expr_list opt_init_expr =
   let size = Array.length all_labels in
-  (* Determine if there are "enough" new fields *)
-  if 3 + 2 * List.length lbl_expr_list >= size
+  (* Determine if there are "enough" fields (only relevant if this is a
+     functional-style record update *)
+  let no_init = match opt_init_expr with None -> true | _ -> false in
+  if no_init || size >= Config.max_young_wosize
```

Comment:
@chambart good catch! Sorry, I did review this but these conditions are trick.

I pushed a fix in 755427318e4ffa1c23b54ab10f11d66e7c4fcf58 , I think it's better to separate concerns and leave your own PR about one thing.